services/subscriptions/subscriptions-supabase.service.ts

```diff
@@ -20,16 +25,79 @@ class SubscriptionSupabase implements SubscriptionService {
-    const { data, error } = await this.supabase.functions.invoke(
-      "stripe-create-subscription"
-    );
-    if (error && error instanceof FunctionsHttpError) {
-      const errorResponse = await error.context.json();
-      messageHandler.handleError(errorResponse.message);
-      return { url: null };
+  public async getSubscriptionLink(): Promise<{ url: string | undefined }> {
+    try {
+      const { data, error } = await this.supabase.functions.invoke<{
+        checkoutUrl: string | undefined;
+      }>("stripe-create-subscription");
+      if (!error && data?.checkoutUrl) {
+        return { url: data?.checkoutUrl }
+      }
+      if (error instanceof FunctionsHttpError) {
+        const errorMessage = await error.context.json();
+        throw errorMessage?.message;
+      } else if (error instanceof FunctionsRelayError) {
+        throw error?.message;
+      } else if (error instanceof FunctionsFetchError) {
+        throw error?.message;
+      }
+      throw "Something went wrong!";
+    } catch (error) {
+      messageHandler.handleError(error?.message || error);
+      return { url: undefined };
+    }
+  }
+
+    /**
+   * retruns the manage subscription link, so that the user can manage their subscription
+   */
+    public async getManageSubscriptionLink(): Promise<{ url: string | undefined }> {
+      try {
+        const { data, error } = await this.supabase.functions.invoke<{
+          manageUrl: string | undefined;
+        }>("stripe-manage-subscription");
+        if (!error && data?.manageUrl) {
+          return { url: data?.manageUrl }
+        }
+        if (error instanceof FunctionsHttpError) {
+          const errorMessage = await error.context.json();
+          throw errorMessage?.message;
+        } else if (error instanceof FunctionsRelayError) {
+          throw error?.message;
+        } else if (error instanceof FunctionsFetchError) {
+          throw error?.message;
+        }
+        throw "Something went wrong!";
+      } catch (error) {
+        messageHandler.handleError(error?.message || error);
+        return { url: undefined };
+      }
+    }
+
+  /**
+   * Cancels the current active subscription and returns the subscription data
+   */
+  public async cancelSubscription(): Promise<Record<string, string | number>> {
+    try {
+      const { data, error } = await this.supabase.functions.invoke(
+        "cancel-subscription"
+      );
+      if (!error) {
+        return data?.data as Record<string, string | number>;
+      }
+      if (error instanceof FunctionsHttpError) {
+        const errorMessage = await error.context.json();
+        throw errorMessage?.message;
+      } else if (error instanceof FunctionsRelayError) {
+        throw error?.message;
+      } else if (error instanceof FunctionsFetchError) {
+        throw error?.message;
+      }
+      throw "Something went wrong!";
+    } catch (error) {
+      messageHandler.handleError(error?.message || error);
+      return {};
```

Comment:
Don't retunr empty object, if an object is expected return null for negative

But here I think just boolean is ok, as we can not expose Supabase objects directly to the frontend (will couple the logic)